tests/copp/test_copp.py

```diff
@@ -0,0 +1,201 @@
+            that have a set rate limit.
+        """
+
+        if protocol == "ARP" \
+                and is_broadcom_device(duthost) \
+                and "201811" not in duthost.get_version():
+            pytest.xfail("ARP policy disabled on BRCM devices due to SAI bug")
+
+        if protocol in ["IP2ME", "SNMP", "SSH"] and _copp_testbed.topo == "t1-lag":
+            pytest.xfail("Packets not received due to faulty DIP, see #1171")
+
+        _copp_runner(duthost,
+                     ptfhost,
+                     protocol,
+                     _copp_testbed)
+
+    @pytest.mark.parametrize("protocol", ["BGP",
+                                          "DHCP",
+                                          "LACP",
+                                          "LLDP",
+                                          "UDLD"])
+    def test_no_policer(self, protocol, duthost, ptfhost, _copp_testbed):
+        """
+            Validates that non-rate-limited COPP groups work as expected.
+
+            Checks that the policer does not enforce a rate limit for protocols
+            that do not have any set rate limit.
+        """
+
+        if protocol == "BGP" and _copp_testbed.topo == "t1-lag":
+            pytest.xfail("Packets not received due to faulty DIP, see #1171")
+
+        _copp_runner(duthost,
+                     ptfhost,
+                     protocol,
+                     _copp_testbed)
+
+@pytest.fixture(scope="class")
+def _copp_testbed(duthost, ptfhost, testbed, request):
+    """
+        Pytest fixture to handle setup and cleanup for the COPP tests.
+    """
+
+    test_params = _gather_test_params(testbed, request)
+
+    if test_params.topo not in _SUPPORTED_TOPOS:
+        pytest.skip("Topology not supported by COPP tests")
+
+    _setup_testbed(duthost, ptfhost, test_params)
+    yield test_params
+    _teardown_testbed(duthost, ptfhost, test_params)
+
+def _copp_runner(dut, ptf, protocol, test_params):
+    """
+        Configures and runs the PTF test cases.
+    """
+
+    params = {"verbose": False,
+              "pkt_tx_count": test_params.pkt_tx_count,
+              "target_port": test_params.nn_target_port}
+
+    dut_ip = dut.setup()["ansible_facts"]["ansible_eth0"]["ipv4"]["address"]
+    device_sockets = ["0-{}@tcp://127.0.0.1:10900".format(test_params.nn_target_port),
+                      "1-{}@tcp://{}:10900".format(test_params.nn_target_port, dut_ip)]
+
+    # NOTE: debug_level can actually slow the PTF down enough to fail the test cases
+    # that are not rate limited. Until this is addressed, do not use this flag as part of
+    # nightly test runs.
+    ptf_runner(host=ptf,
+               testdir="ptftests",
+               testname="copp_tests.{}Test".format(protocol),
+               platform="nn",
+               qlen=100000,
+               params=params,
+               relax=None,
+               debug_level=None,
+               device_sockets=device_sockets)
```

Comment:
What is the reason to use ptf_runner instead of ptfadapter? 
With ptfadapter we can do the same what ptftest, but directly from mgmt container from pytest. And this is really helpful in case of test debugging.

Reply:
I used the ptf_runner to keep the size of this PR reasonable. I want to use ptfadapter for the reasons you mentioned, but the COPP ptf script is pretty big so I want to do that conversion in a follow-up PR.